File: app/src/main/java/com/helpcrunch/helpcrunchdemo/screens/CustomUserDataActivity.java
package com.helpcrunch.helpcrunchdemo.screens;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.helpcrunch.helpcrunchdemo.R;
import com.helpcrunch.library.core.Callback;
import com.helpcrunch.library.core.HelpCrunch;
import com.helpcrunch.library.core.models.user.HCUser;

import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;

public class CustomUserDataActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_custom_user_data);

        initViews();

        fillData();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_custom_data, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            this.finish();
            return true;
        } else if (item.getItemId() == R.id.add_row) {
            addRow(null, null);
        }
        return super.onOptionsItemSelected(item);
    }

    private void initViews() {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(R.string.change_custom_user_data);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        findViewById(R.id.save_user_data_button).setOnClickListener(view -> {
            saveCustomData();
        });
    }

    private void fillData() {
        HCUser currentUser = HelpCrunch.getUser();

        if (currentUser == null) return;

        Map<String, Object> data = currentUser.getCustomData();

        if (data != null) {
            for (Map.Entry<String, Object> entry : data.entrySet()) {
                addRow(entry.getKey(), entry.getValue().toString());
            }
        }
    }

    private void saveCustomData() {
        if (HelpCrunch.getUser() == null) {
            Toast.makeText(this, "There is no user yet", Toast.LENGTH_SHORT).show();
            return;
        }

        HashMap<String, Object> customData = new HashMap<>();
        LinearLayout form = findViewById(R.id.registration_form);

        getNewData(customData, form);

        HCUser user = HelpCrunch.getUser();
        user.setCustomData(customData);

        setUserDataButtonParameters(View.VISIBLE, false);

        HelpCrunch.updateUser(user, new Callback<HCUser>() {
            @Override
            public void onSuccess(HCUser result) {
                Toast.makeText(CustomUserDataActivity.this, getString(R.string.data_saved), Toast.LENGTH_SHORT).show();
                setUserDataButtonParameters(View.GONE, true);
            }

            @Override
            public void onError(@NotNull String message) {
                Toast.makeText(CustomUserDataActivity.this, getString(R.string.something_wrong), Toast.LENGTH_SHORT).show();
                setUserDataButtonParameters(View.GONE, true);
            }
        });
    }

    private void getNewData(HashMap<String, Object> customData, LinearLayout form) {
        for (int i = 0; i < form.getChildCount(); i++) {
            View child = form.getChildAt(i);

            if (child instanceof LinearLayout) {
                String key = ((EditText) child.findViewById(R.id.key)).getText().toString();
                String value = ((EditText) child.findViewById(R.id.value)).getText().toString();

                addData(customData, key, value);
            }
        }
    }

    private void addRow(String key, String value) {
        LinearLayout form = findViewById(R.id.registration_form);

        View view = LayoutInflater.from(this).inflate(R.layout.layout_custom_user_data_row, null);

        ((EditText) view.findViewById(R.id.key)).setText(key);
        ((EditText) view.findViewById(R.id.value)).setText(value);

        form.addView(view);
    }

    private void addData(Map<String, Object> customData, String key, String object) {
        if (!TextUtils.isEmpty(key) && !TextUtils.isEmpty(object)) {
            customData.put(key, object);
        }
    }

    private void setUserDataButtonParameters(int visible, boolean enabled) {
        findViewById(R.id.save_user_data_button_progress).setVisibility(visible);
        findViewById(R.id.save_user_data_button).setEnabled(enabled);
    }
}

File: app/src/main/java/com/helpcrunch/demo/screens/CustomUserDataActivity.kt
package com.helpcrunch.demo.screens

import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import com.helpcrunch.demo.R
import com.helpcrunch.demo.databinding.ActivityCustomUserDataBinding
import com.helpcrunch.library.core.Callback
import com.helpcrunch.library.core.HelpCrunch.getUser
import com.helpcrunch.library.core.HelpCrunch.updateUser
import com.helpcrunch.library.core.models.user.HCUser

class CustomUserDataActivity : AppCompatActivity() {
    private lateinit var binding: ActivityCustomUserDataBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityCustomUserDataBinding.inflate(layoutInflater)

        setContentView(binding.root)

        initViews()
        fillData()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun initViews() {
        ViewCompat.setOnApplyWindowInsetsListener(binding.root) { v, windowInsets ->
            val insets = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars())

            v.setPadding(
                insets.left,
                insets.top,
                insets.right,
                insets.bottom
            )

            WindowInsetsCompat.CONSUMED
        }

        WindowCompat.getInsetsController(window, window.decorView)
            .isAppearanceLightStatusBars = true

        setSupportActionBar(binding.toolbar)

        if (supportActionBar != null) {
            supportActionBar!!.setTitle(R.string.send_message)
            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        }

        binding.saveUserDataButton.setOnClickListener { saveCustomData() }
        binding.addNewItem.setOnClickListener { addRow(null, null) }
    }

    private fun fillData() {
        val data = getUser()?.customData ?: return

        if (data.isNotEmpty()) {
            for ((key, value) in data) {
                addRow(key, value.toString())
            }
        } else {
            addRow(null, null)
        }
    }

    private fun saveCustomData() {
        binding.saveUserDataButtonIcon.setLoading(true)
        binding.saveUserDataButton.isEnabled = false

        val customData = HashMap<String, Any?>()
        getNewData(customData, binding.registrationForm)
        val user = HCUser.Builder(getUser())
            .withCustomData(customData)
            .build()
        updateUser(user, object : Callback<HCUser>() {
            override fun onSuccess(result: HCUser) {
                Toast.makeText(
                    this@CustomUserDataActivity,
                    getString(R.string.data_saved),
                    Toast.LENGTH_SHORT
                ).show()

                binding.saveUserDataButtonIcon.setLoading(false)
                binding.saveUserDataButton.isEnabled = true
            }

            override fun onError(message: String) {
                Toast.makeText(
                    this@CustomUserDataActivity,
                    getString(R.string.something_wrong),
                    Toast.LENGTH_SHORT
                ).show()

                binding.saveUserDataButtonIcon.setLoading(false)
                binding.saveUserDataButton.isEnabled = true
            }
        })
    }

    private fun getNewData(customData: HashMap<String, Any?>, form: LinearLayout) {
        for (i in 0 until form.childCount) {
            val child = form.getChildAt(i)
            if (child is LinearLayout) {
                val key = (child.findViewById<View>(R.id.key) as EditText).text.toString()
                val value = (child.findViewById<View>(R.id.value) as EditText).text.toString()
                addData(customData, key, value)
            }
        }
    }

    private fun addRow(key: String?, value: String?) {
        val view = LayoutInflater.from(this).inflate(R.layout.layout_custom_user_data_row, null)
        (view.findViewById<View>(R.id.key) as EditText).setText(key)
        (view.findViewById<View>(R.id.value) as EditText).setText(value)
        binding.registrationForm.addView(view)
    }

    private fun addData(customData: MutableMap<String, Any?>, key: String, item: String) {
        if (!TextUtils.isEmpty(key) && !TextUtils.isEmpty(item)) {
            customData[key] = item
        }
    }
}
